src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractSubmitToArchiveCommand.java

```diff
@@ -0,0 +1,99 @@
+import edu.harvard.iq.dataverse.settings.SettingsServiceBean;
+import edu.harvard.iq.dataverse.workflow.step.WorkflowStepResult;
+
+import static edu.harvard.iq.dataverse.engine.command.CommandHelper.CH;
+
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+
+@RequiredPermissions(Permission.PublishDataset)
+public abstract class AbstractSubmitToArchiveCommand implements Command<DatasetVersion> {
+
+    private final DatasetVersion version;
+    private final DataverseRequest request;
+    private final Map<String, DvObject> affectedDvObjects;
+    private final Map<String, String> requestedSettings = new HashMap<String, String>();
+    private static final Logger logger = Logger.getLogger(AbstractSubmitToArchiveCommand.class.getName());
+    
+    public AbstractSubmitToArchiveCommand(DataverseRequest aRequest, DatasetVersion version) {
+        this.request = aRequest;
+        this.version = version;
+        affectedDvObjects = new HashMap<>();
+        affectedDvObjects.put("", version.getDataset());
+    }
+
+    @Override
+    public DatasetVersion execute(CommandContext ctxt) throws CommandException {
+
+        String settings = ctxt.settings().getValueForKey(SettingsServiceBean.Key.ArchiverSettings);
+        String[] settingsArray = settings.split(",");
+        for (String setting : settingsArray) {
+            setting = setting.trim();
+            if (!setting.startsWith(":")) {
+                logger.warning("Invalid Archiver Setting: " + setting);
+            } else {
+                requestedSettings.put(setting, ctxt.settings().get(setting));
+            }
+        }
+        
+        AuthenticatedUser user = request.getAuthenticatedUser();
+        ApiToken token = ctxt.authentication().findApiTokenByUser(user);
+        if ((token == null) || (token.getExpireTime().before(new Date()))) {
+            token = ctxt.authentication().generateApiTokenForUser(user);
+        }
+        performArchiveSubmission(version, token, requestedSettings);
+        return ctxt.em().merge(version);
+    }
+
+    @Override
+    public Map<String, DvObject> getAffectedDvObjects() {
+
+        return affectedDvObjects;
+    }
+
+    @Override
+    public DataverseRequest getRequest() {
+        return request;
+    }
+
+    @Override
+    public Map<String, Set<Permission>> getRequiredPermissions() {
+        return CH.permissionsRequired(getClass());
+    }
+
+    @Override
+    public String describe() {
+        // Is this ever used?
+        return "DatasetVersion: " + version.getId() + " " + version.getDataset().getDisplayName() + ".v"
+                + version.getFriendlyVersionNumber();
+    }
+
+    /**
+     * This method is the only one that should be overwritten by other classes. Note
+     * that this method may be called from the execute method above OR from a
+     * workflow in which execute() is never called and therefore in which all
```

Comment:
I haven't looked in workflows in detail.Is there a reason they can't just call execute?

Reply:
Definitely some room for redesign (I also consulted with @michbarsinai who had worked on both these sections - " we could refactor the Commands and WorkFlow packages to be a bit closer at some point. Or maybe add a workflow step wrapper for commands or something". But we'll leave that for maintenance and out of scope for this issue.